From 1b8cc5047462ce0dce53bab0bf5b94eec2b5f9db Mon Sep 17 00:00:00 2001
From: Robert S Peterson <rpeterso@redhat.com>
Date: Tue, 11 Nov 2014 18:10:25 -0500
Subject: [fs] GFS2: Only increase rs_sizehint

Message-id: <b8132be6f72425050524f5172d12f526c7848a95.1415729212.git.rpeterso@redhat.com>
Patchwork-id: 99828
O-Subject: [RHEL7.1 PATCH fs 2/3] GFS2: Only increase rs_sizehint
Bugzilla: 1142238
RH-Acked-by: Abhijith Das <adas@redhat.com>
RH-Acked-by: Steven Whitehouse <swhiteho@redhat.com>

If an application does a sequence of (1) big write, (2) little write
we don't necessarily want to reset the size hint based on the smaller
size. The fact that they did any big writes implies they may do more,
and therefore we should try to allocate bigger block reservations, even
if the last few were small writes. Therefore this patch changes function
gfs2_size_hint so that the size hint can only grow; it cannot shrink.
This is especially important where there are multiple writers.

rhbz#1142238
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/gfs2/file.c b/fs/gfs2/file.c
index c301812..6afdcc9 100644
--- a/fs/gfs2/file.c
+++ b/fs/gfs2/file.c
@@ -344,7 +344,8 @@ static void gfs2_size_hint(struct file *filep, loff_t offset, size_t size)
  size_t blks = (size + sdp->sd_sb.sb_bsize - 1) >> sdp->sd_sb.sb_bsize_shift;
  int hint = min_t(size_t, INT_MAX, blks);
 
- atomic_set(&ip->i_res->rs_sizehint, hint);
+ if (hint > atomic_read(&ip->i_res->rs_sizehint))
+  atomic_set(&ip->i_res->rs_sizehint, hint);
 }
 
 /**
-- 
1.7.1